xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -125,9 +121,163 @@ private static Bootstrap readConfig() throws IOException {
   }
 
   @VisibleForTesting
-  static Bootstrap parseConfig(String rawData) throws InvalidProtocolBufferException {
-    Bootstrap.Builder bootstrapBuilder = Bootstrap.newBuilder();
-    JsonFormat.parser().merge(rawData, bootstrapBuilder);
-    return bootstrapBuilder.build();
+  @SuppressWarnings("unchecked")
+  static BootstrapInfo parseConfig(String rawData) throws IOException {
+    Map<String, ?> rawBootstrap = (Map<String, ?>) JsonParser.parse(rawData);
+
+    Map<String, ?> rawServerConfig = JsonUtil.getObject(rawBootstrap, "xds_server");
+    if (rawServerConfig == null) {
+      throw new IOException("Invalid bootstrap: 'xds_server' does not exist.");
+    }
+    // Field "server_uri" is required.
+    String serverUri = JsonUtil.getString(rawServerConfig, "server_uri");
+    if (serverUri == null) {
+      throw new IOException("Invalid bootstrap: 'xds_server : server_uri' does not exist.");
+    }
+    List<ChannelCreds> channelCredsOptions = new ArrayList<>();
+    List<?> rawChannelCredsList = JsonUtil.getList(rawServerConfig, "channel_creds");
+    // List of channel creds is optional.
+    if (rawChannelCredsList != null) {
+      for (Object channelCreds : rawChannelCredsList) {
+        Map<String, ?> channelCredsItem = (Map<String, ?>) channelCreds;
+        String type = (String) JsonUtil.getString(channelCredsItem, "type");
+        if (type == null) {
+          throw new IOException("Invalid bootstrap: 'channel_creds' contains unknown type.");
+        }
+        ChannelCreds creds = new ChannelCreds(type);
+        if (channelCredsItem.containsKey("config")) {
+          creds.config = JsonUtil.getObject(channelCredsItem, "config");
```

Comment:
Why not make `config` a constructor arg and final field of `ChannelCreds` as well?

Reply:
It looks bad to have `new ChannelCreds("TLS", null)` (although it's only in test). We might consider doing that later if the structure of `config` gets more complex. But for now, I believe it's not a big deal.

Reply:
It looks worse to have both a final and a non-final field in data class.

Reply:
Plus, `JsonUtil.getObject()` is `Nullable`, so you don't explicitly call `new ChannelCreds(type, null)` in the main class. Where do you have to call `new ChannelCreds("TLS", null)` in test class?

Reply:
Because how exactly `config` would look like is still undefined.

Reply:
Changed `config` field to be final.